tf_helpers/tf_jax_stax.py

```diff
@@ -0,0 +1,338 @@
+    kernel_shape = [out_chan if c == 'O' else
+                    input_shape[lhs_spec.index('C')] if c == 'I' else
+                    next(filter_shape_iter) for c in rhs_spec]
+    output_shape = lax.conv_general_shape_tuple(
+        input_shape, kernel_shape, strides, padding, dimension_numbers)
+    bias_shape = [out_chan if c == 'C' else 1 for c in out_spec]
+    bias_shape = tuple(itertools.dropwhile(lambda x: x == 1, bias_shape))
+    keys = split(seed=tf.convert_to_tensor(rng, dtype=tf.int32), num=2)
+    k1 = keys[0]
+    k2 = keys[1]
+    W = W_init(seed=k1, shape=kernel_shape)
+    b = b_init(stddev=1e-6, seed=k2, shape=bias_shape)
+    return tfnp.zeros(output_shape), (W, b)
+  def apply_fun(params, inputs, **kwargs):
+    W, b = params
+    return lax.conv_general_dilated(inputs, W, strides, padding, one, one,
+                                    dimension_numbers=dimension_numbers) + b
+  return init_fun, apply_fun
+Conv = functools.partial(GeneralConv, ('NHWC', 'HWIO', 'NHWC'))
+
+
+def elementwise(fun, **fun_kwargs):
+  """Layer that applies a scalar function elementwise on its inputs."""
+  def init_fun(rng, input_shape):
+    return (tfnp.zeros(input_shape), ())
+  # init_fun = lambda rng, input_shape: (tfnp.zeros(input_shape), ())
+  apply_fun = lambda params, inputs, **kwargs: fun(inputs, **fun_kwargs)
+  return init_fun, apply_fun
+Tanh = elementwise(tfnp.tanh)
+Relu = elementwise(relu)
+Exp = elementwise(tfnp.exp)
+LogSoftmax = elementwise(log_softmax, axis=-1)
+Softmax = elementwise(softmax, axis=-1)
+Softplus = elementwise(softplus)
+Sigmoid = elementwise(sigmoid)
+Elu = elementwise(elu)
+LeakyRelu = elementwise(leaky_relu)
+Selu = elementwise(selu)
+
+
+def _pooling_layer(reducer, init_val, rescaler=None):
+  def PoolingLayer(window_shape, strides=None, padding='VALID', spec=None):
+    """Layer construction function for a pooling layer."""
+    strides = strides or (1,) * len(window_shape)
+    rescale = rescaler(window_shape, strides, padding) if rescaler else None
+
+    dim = len(window_shape)
+    batch_dim, channel_dim = None, None
+    if spec is None:
+      batch_dim, channel_dim = 0, len(window_shape) + 1
+    else:
+      batch_dim, channel_dim = spec.index('N'), spec.index('C')
+    window_shape = window_shape
+    strides = strides
+
+    def init_fun(rng, input_shape):
+      # Move the batch and channel dimension of the input shape such
+      # that it is of data format "NHWC"
+      shape = [input_shape[batch_dim]]
+      for i in range(len(input_shape)):
+        if i not in [batch_dim, channel_dim]:
+          shape.append(input_shape[i])
+      shape.append(input_shape[channel_dim])
+      out_shape = lax.reduce_window_shape_tuple(shape, window_shape,
+                                                strides, padding)
+      return tfnp.zeros(out_shape), ()
+    def apply_fun(params, inputs, **kwargs):
+      inputs = onp.moveaxis(inputs, (batch_dim, channel_dim), \
```

Comment:
Would the use of `onp` here transfer `inputs` from GPU/TPU to CPU? Should we try to avoid this?

Also, I believe this is the kind of transposition that can be avoided by using the default `lax` API.